### PR TITLE
[v7] Update `BTThreeDSecureRequest` Parameters

### PR DESCRIPTION
**Pull request by agedd**
### Summary of changes

- Update `BTThreeDSecureRequest` properties to have everything in the init vs using dot syntax
- Update related unit tests 
- Confirmed with the gateway on questions around required vs optional parameters

### Checklist

- [x] Added a changelog entry

- @agedd

### CODE REVIEW

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 1]
```diff
@@ -7,97 +7,89 @@ import BraintreeCore
...
+        email: String? = nil,
+        exemptionRequested: Bool = false,
+        mobilePhoneNumber: String? = nil,
+        nonce: String? = nil,
+        renderTypes: [BTThreeDSecureRenderType]? = nil,
+        requestedExemptionType: BTThreeDSecureRequestedExemptionType = .unspecified,
+        shippingMethod: BTThreeDSecureShippingMethod = .unspecified,
+        uiType: BTThreeDSecureUIType = .both,
+        v2UICustomization: BTThreeDSecureV2UICustomization? = nil
+    ) {
```
Comment (agedd) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:75]: note: alphabetized the order of the parameters in the initializer

File: Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift [hunk 2]
```diff
@@ -424,9 +424,9 @@ import BraintreeCore
             requestParameters["customFields"] = customFields
         }
 
-        if request._cardAddChallenge == .requested || request.cardAddChallengeRequested == true {
+        if request.cardAddChallengeRequested == true {
             requestParameters["cardAdd"] = true
-        } else if request._cardAddChallenge == .notRequested {
+        } else if request.cardAddChallengeRequested == false {
```
Comment (agedd) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:429]: question for the team: does this logic sound okay? (updated this after removing the `cardAddChallenge` property based on its doc strings)

Comment (jaxdesmarais) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:429]: There is a separate ticket for this: DTMOBILES-1140. So we can only do the init portion in this PR.

Comment (agedd) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:429]: ahh gotcha! thanks, addressed in b2bec4a7a7638818e90bb3b13d658912b0b9a9a4

Comment (jaxdesmarais) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:429]: I think we can revert the changes here and uncomment `_cardAddChallenge` for now, then we can tackle all of it in a single PR.

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift (hunk 1 above)
Comment (warmkesselj) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:95]: What do you think about grouping the required properties first and then the non-optional properties?

**Review by warmkesselj — APPROVED** [review]
One comment but it's not a blocker. Approving ahead of time.

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 3]
```diff
@@ -66,38 +42,74 @@ import BraintreeCore
...
-    // MARK: - Internal Properties
+    // MARK: - Initializer
     
-    /// The dfReferenceID for the session. Exposed for testing.
-    var dfReferenceID: String?
+    /// Creates a `BTThreeDSecureRequest`
+    /// - Parameters:
+    ///    - accountType: Optional. The account type selected by the cardholder. Some cards can be processed using either a credit or debit account and cardholders have the option to choose which account to use.
+    ///    - additionalInformation: Optional. The additional information used for verification.
+    ///    - amount: The amount for the transaction.
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:60]: If this is optional do we want to note that in the docstring?

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 4]
```diff
@@ -66,38 +42,74 @@ import BraintreeCore
...
+    ///    - billingAddress: Optional. The billing address used for verification
+    ///    - cardAddChallengeRequested: Optional.  An authentication created using this flag should only be used for vaulting operations (creation of customers' credit cards or payment methods) and not for creating transactions. If set to `true`, a card-add challenge will be requested from the issuer. If set to `false`, a card-add challenge will not be requested. If the parameter is missing, a card-add challenge will only be requested for $0 amount.
+    ///    - challengeRequested: Optional. If set to true, an authentication challenge will be forced if possible.
+    ///    - customFields: Object where each key is the name of a custom field which has been configured in the Control Panel. In the Control Panel you can configure 3D Secure Rules which trigger on certain values.
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:64]: Same here, do we want to note this as optional in the docstring?

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 5]
```diff
@@ -66,38 +42,74 @@ import BraintreeCore
...
+    ///    - dataOnlyRequested:  Optional. Indicates whether to use the data only flow. In this flow, frictionless 3DS is ensured for Mastercard cardholders as the card scheme provides a risk score for the issuer to determine whether to approve. If data only is not supported by the processor, a validation error will be raised. Non-Mastercard cardholders will fallback to a normal 3DS flow.
+    ///    - dfReferenceID: The dfReferenceID for the session. Exposed for testing.
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:66]: Should this be exposed for merchants? It looks like it was internal previously.

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 6]
```diff
@@ -66,38 +42,74 @@ import BraintreeCore
...
+    ///    - email: Optional. The email used for verification.
+    ///    - exemptionRequested: Optional. If set to true, an exemption to the authentication challenge will be requested.
+    ///    - mobilePhoneNumber: Optional. The mobile phone number used for verification. Only numbers. Remove dashes, parentheses and other characters.
+    ///    - nonce: A nonce to be verified by ThreeDSecure
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:70]: Do we want to add optional to the docstring here if this is optional?

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 7]
```diff
@@ -66,38 +42,74 @@ import BraintreeCore
...
+    ///    - renderTypes: Optional: List of all the render types that the device supports for displaying specific challenge user interfaces within the 3D Secure challenge. When using `BTThreeDSecureUIType.both` or `BTThreeDSecureUIType.html`, all `BTThreeDSecureRenderType` options must be set. When using `BTThreeDSecureUIType.native`, all `BTThreeDSecureRenderType` options except `.html` must be set.
+    ///    - requestedExemptionType: Optional. The exemption type to be requested. If an exemption is requested and the exemption's conditions are satisfied, then it will be applied.
+    ///    - shippingMethod: Optional. The shipping method chosen for the transaction
+    ///    - uiType: Optional: Sets all UI types that the device supports for displaying specific challenge user interfaces in the 3D Secure challenge. Defaults to `.both`
+    ///    - v2UICustomization: Optional. UI Customization for 3DS2 challenge views.
+    public init(
+        accountType: BTThreeDSecureAccountType = .unspecified,
+        additionalInformation: BTThreeDSecureAdditionalInformation? = nil,
+        amount: NSDecimalNumber? = 0,
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:79]: Based on this logic it looks like both amount and the delegate may be required: https://github.com/braintree/braintree_ios/blob/5ecb18dc594019e5bf9292fd67871b8f15812fd1/Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift#L62-L74

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift (hunk 5 above)
Comment (jaxdesmarais) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:66]: Ah, I see it looks like this is exposed for merchants using `prepareLookup`: https://github.com/braintree/braintree_ios/blob/5ecb18dc594019e5bf9292fd67871b8f15812fd1/Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift#L126-L128. We should update this docstring or have 2 inits for this. What do you think?

Comment (agedd) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift]: ahhh that's a good point! i've updated the docstrings to say, that the `dfReferenceID` is helpful for merchants using the 3DS `prepareLookup` in b6f5382aefb7dc7a3467a891735988bba0a80a0d

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift (hunk 7 above)
Comment (agedd) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift]: ahh yep, good catch! the amount does seem to be required (i've also confirmed this by raising a BT 3D Secure General Inquiry workflow request) - as for the delegate, it seems that it was also previously optional so i wonder if it's look to leave that as is 👀

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,12 +6,14 @@
...
-  * BraintreeLocalPayment
-    * Update `BTLocalPaymentRequest` to make all properties accessible on the initializer only vs via the dot syntax.
+* BraintreeVenmo
+  * Update `BTVenmoRequest` to make all properties accessible on the initializer only vs via the dot syntax.
+* BraintreeSEPADirectDebit
+  * Update `BTSEPADirectDebitRequest` to make all properties accessible on the initializer only vs via the dot syntax.
+* BraintreeLocalPayment
+  * Update `BTLocalPaymentRequest` to make all properties accessible on the initializer only vs via the dot syntax.
+* BraintreeThreeDSecure
+  * Update `BraintreeThreeDSecure` to make all properties accessible on the initializer only vs via the dot syntax.
```
Comment (jaxdesmarais) [inline CHANGELOG.md:16]: Lets keep the above spacing. These are breaking changes so we want them to be nested under the breaking changes section. See here for similar formatting with v6: https://github.com/braintree/braintree_ios/blob/main/CHANGELOG.md#600-beta4-2023-06-01

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 9]
```diff
@@ -66,38 +42,76 @@ import BraintreeCore
...
+    ///    - challengeRequested: Optional. If set to true, an authentication challenge will be forced if possible.
+    ///    - customFields: Optional. Object where each key is the name of a custom field which has been configured in the Control Panel. In the Control Panel you can configure 3D Secure Rules which trigger on certain values.
+    ///    - dataOnlyRequested:  Optional. Indicates whether to use the data only flow. In this flow, frictionless 3DS is ensured for Mastercard cardholders as the card scheme provides a risk score for the issuer to determine whether to approve. If data only is not supported by the processor, a validation error will be raised. Non-Mastercard cardholders will fallback to a normal 3DS flow.
+    ///    - dfReferenceID: Optional. The dfReferenceID for the session, particularly useful for merchants peforming 3DS lookup. Exposed for testing.
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:66]: Let's remove the testing comment from the public docs.
```suggestion
    ///    - dfReferenceID: Optional. The dfReferenceID for the session, particularly useful for merchants peforming 3DS lookup.
```

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 10]
```diff
@@ -66,38 +42,76 @@ import BraintreeCore
...
-    // MARK: - Internal Properties
+    // MARK: - Initializer
     
-    /// The dfReferenceID for the session. Exposed for testing.
-    var dfReferenceID: String?
+    /// Creates a `BTThreeDSecureRequest`
+    /// - Parameters:
+    ///    - accountType: Optional. The account type selected by the cardholder. Some cards can be processed using either a credit or debit account and cardholders have the option to choose which account to use.
+    ///    - additionalInformation: Optional. The additional information used for verification.
+    ///    - amount: Required. The amount for the transaction.
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:60]: If this is required lets put it first and not provide a default. It's also worth considering if we want to continue to use `NSDecimalNumber` or if we can switch to `Double`?

Comment (agedd) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift]: ahh yep, so i just verified with the public docs and it looks like `amount` is a `string` and required (same with the `nonce` property) - updated the initializer and other places in the code accordingly

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 11]
```diff
@@ -66,38 +42,76 @@ import BraintreeCore
...
+    ///    - shippingMethod: Optional. The shipping method chosen for the transaction
+    ///    - uiType: Optional: Sets all UI types that the device supports for displaying specific challenge user interfaces in the 3D Secure challenge. Defaults to `.both`
+    ///    - v2UICustomization: Optional. UI Customization for 3DS2 challenge views.
+    public init(
+        amount: String,
+        nonce: String,
+        accountType: BTThreeDSecureAccountType = .unspecified,
+        additionalInformation: BTThreeDSecureAdditionalInformation? = nil,
+        billingAddress: BTThreeDSecurePostalAddress? = nil,
+        _cardAddChallenge: BTThreeDSecureCardAddChallenge = .unspecified,
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:82]: nit: we can probably just leave this out of the init for now then remove it in the ticket to remove the property completely. If that's more work than it's worth though, feel free to leave and we can hopefully pick up that cleanup ticket soon!

File: Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift [hunk 12]
```diff
@@ -66,38 +42,76 @@ import BraintreeCore
...
+    ///    - mobilePhoneNumber: Optional. The mobile phone number used for verification. Only numbers. Remove dashes, parentheses and other characters.
+    ///    - renderTypes: Optional: List of all the render types that the device supports for displaying specific challenge user interfaces within the 3D Secure challenge. When using `BTThreeDSecureUIType.both` or `BTThreeDSecureUIType.html`, all `BTThreeDSecureRenderType` options must be set. When using `BTThreeDSecureUIType.native`, all `BTThreeDSecureRenderType` options except `.html` must be set.
+    ///    - requestedExemptionType: Optional. The exemption type to be requested. If an exemption is requested and the exemption's conditions are satisfied, then it will be applied.
+    ///    - shippingMethod: Optional. The shipping method chosen for the transaction
+    ///    - uiType: Optional: Sets all UI types that the device supports for displaying specific challenge user interfaces in the 3D Secure challenge. Defaults to `.both`
+    ///    - v2UICustomization: Optional. UI Customization for 3DS2 challenge views.
+    public init(
+        amount: String,
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:77]: 🤔  I am getting an error with the amount as a string in the demo app. Are we sure this is valid? It looks like a `Double` works as expected.

Comment (agedd) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:77]: hmm yea i'm a bit confused on this - i also feel amount as `Double` makes more sense (and this aligns with what is mentioned in our iOS BT Dev Docs) but after asking here  #[bt-service-3dsecure](https://paypal.slack.com/archives/C02743S2L/p1731618309941569) and checking the android [equivalent](https://github.com/braintree/braintree_android/blob/c9d4246eccb6926a0985b5e6d409147d6fb38470/ThreeDSecure/src/main/java/com/braintreepayments/api/threedsecure/ThreeDSecureRequest.kt#L56) it looks like they define amount as a string

Comment (jaxdesmarais) [reply Sources/BraintreeThreeDSecure/BTThreeDSecureRequest.swift:77]: Ah, you're right, I must have been doing something weird the first time I tested it as a string on main that resulted in an error. If we want to follow what android/the service channel recommends that's okay with me.

File: Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift [hunk 13]
```diff
@@ -59,7 +59,7 @@ import BraintreeCore
                 return
             }
 
-            if request.amount?.decimalValue.isNaN == true || request.amount == nil {
+            if request.amount.isEmpty {
                 NSLog("%@ BTThreeDSecureRequest amount can not be nil or NaN.", BTLogLevelDescription.string(for: .critical))
                 let error = BTThreeDSecureError.configuration("BTThreeDSecureRequest amount can not be nil or NaN.")
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:64]: Lets update these strings to accurately reflect this block now that amount cannot be nil or NaN. We can instead update to something like "BTThreeDSecureRequest amount cannot be an empty string"

File: Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift [hunk 14]
```diff
@@ -103,7 +103,7 @@ import BraintreeCore
             return
         }
 
-        guard request.nonce != nil else {
+        if request.nonce.isEmpty {
             notifyFailure(
                 with: BTThreeDSecureError.configuration("BTThreeDSecureRequest nonce can not be nil."),
```
Comment (jaxdesmarais) [inline Sources/BraintreeThreeDSecure/BTThreeDSecureClient.swift:108]: Same here, can also update to something like "BTThreeDSecureRequest nonce can not be an empty string"

File: CHANGELOG.md [hunk 15]
```diff
@@ -15,6 +15,8 @@
     * Update `BTLocalPaymentRequest` to make all properties accessible on the initializer only vs via the dot syntax.
   * BraintreeCard 
     * Remove `BTCardRequest`, use `BTCard` directly instead
+  * BraintreeThreeDSecure
+    * Update `BTThreeDSecureRequest` to make all properties accessible on the initializer only vs via the dot syntax.
```
Comment (jaxdesmarais) [inline CHANGELOG.md:19]: Let's add an entry as well for any change in types like `BTThreeDSecureRequest.amount` now being a string

**Pull request merged**